`redundant_discardable_let` rule causes a compilation error for things that are wrapped in a result-builder annotation

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`redundant_discardable_let` rule causes a compilation error for things that are wrapped in a result-builder annotation.

Here's a link to the rule on GitHub which contains the bug: https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/Style/RedundantDiscardableLetRule.swift

It's unsafe, for say, the `body` property of a struct which conforms to the `View` protocol from SwiftUI (and which therefore has a `body` property that's implicitly-annotated with the `@ViewBuilder` result-builder annotation), to replace this code snippet:
```swift
// Inside struct conforming to `View` protocol from `SwiftUI`.
var body: some View {
    let _ = 10

    Text("Hello, world!")
        .padding()
}
```

with this code snippet:
```swift
// Inside struct conforming to `View` protocol from `SwiftUI`.
var body: some View {
    _ = 10

    Text("Hello, world!")
        .padding()
}
```
as making such a replacement, would result in a compilation error. In this specific example, that compilation error will be: "Type '()' cannot conform to 'View'".

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint /Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift
Linting Swift files at paths /Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift
Linting 'ContentView.swift' (1/1)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:13:9: warning: Redundant Discardable Let Violation: Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. (redundant_discardable_let)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environments

* SwiftLint version: 0.46.2
* Installation method used: Homebrew
* Paste your configuration file:
N/A - default settings were used.

* Nested configurations are not being used.
* Xcode version: Xcode 13.2 Build version 13C5066c
* Running the command: `swiftlint lint --path /Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift --no-cache --enable-all-rules` produced the following output:
```bash
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:12:5: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:11:1: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:21:5: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:20:1: warning: Explicit ACL Violation: All declarations should specify Access Control Level keywords explicitly. (explicit_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:11:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly. (explicit_top_level_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:20:1: warning: Explicit Top Level ACL Violation: Top-level declarations should specify Access Control Level keywords explicitly. (explicit_top_level_acl)
/Users/jeremy/dev/xcode/SwiftLintBugDemoApp/SwiftLintBugDemoApp/ContentView.swift:13:9: warning: Redundant Discardable Let Violation: Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. (redundant_discardable_let)
Done linting! Found 7 violations, 0 serious in 1 file.
```

```swift
import SwiftUI

struct ContentView: View {
    var body: some View {
        let _ = 10 // This line triggers a violation.

        Text("Hello, world!")
            .padding()
    }
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -43,6 +43,10 @@
   [dk-talks](https://github.com/dk-talks)
   [SimplyDanny](https://github.com/SimplyDanny)
 
+* Add option to disable `redundant_discardable_let` rule in SwiftUI view bodies.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#3855](https://github.com/realm/SwiftLint/issues/3855)
+
 * Add new `redundant_sendable` rule that triggers on `Sendable` conformances of
   types that are implicitly already `Sendable` due to being actor-isolated. It
   is enabled by default.  

--- a/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/RedundantDiscardableLetConfiguration.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/RedundantDiscardableLetConfiguration.swift
@@ -1,0 +1,11 @@
+import SwiftLintCore
+
+@AutoConfigParser
+struct RedundantDiscardableLetConfiguration: SeverityBasedRuleConfiguration {
+    typealias Parent = RedundantDiscardableLetRule
+
+    @ConfigurationElement(key: "severity")
+    private(set) var severityConfiguration = SeverityConfiguration<Parent>(.warning)
+    @ConfigurationElement(key: "ignore_swiftui_view_bodies")
+    private(set) var ignoreSwiftUIViewBodies = false
+}

--- a/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift
@@ -1,8 +1,8 @@
 import SwiftSyntax
 
-@SwiftSyntaxRule(explicitRewriter: true)
+@SwiftSyntaxRule(correctable: true)
 struct RedundantDiscardableLetRule: Rule {
-    var configuration = SeverityConfiguration<Self>(.warning)
+    var configuration = RedundantDiscardableLetConfiguration()
 
     static let description = RuleDescription(
         identifier: "redundant_discardable_let",
@@ -16,10 +16,22 @@ struct RedundantDiscardableLetRule: Rule {
             Example("let _: ExplicitType = foo()"),
             Example("while let _ = SplashStyle(rawValue: maxValue) { maxValue += 1 }"),
             Example("async let _ = await foo()"),
+            Example("""
+                var body: some View {
+                    let _ = foo()
+                    return Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true]),
         ],
         triggeringExamples: [
             Example("↓let _ = foo()"),
             Example("if _ = foo() { ↓let _ = bar() }"),
+            Example("""
+                var body: some View {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """),
         ],
         corrections: [
             Example("↓let _ = foo()"): Example("_ = foo()"),
@@ -29,35 +41,63 @@ struct RedundantDiscardableLetRule: Rule {
 }
 
 private extension RedundantDiscardableLetRule {
-    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
-        override func visitPost(_ node: VariableDeclSyntax) {
-            if node.hasRedundantDiscardableLetViolation {
-                violations.append(node.positionAfterSkippingLeadingTrivia)
-            }
-        }
+    private enum CodeBlockKind {
+        case normal
+        case view
     }
 
-    final class Rewriter: ViolationsSyntaxRewriter<ConfigurationType> {
-        override func visit(_ node: VariableDeclSyntax) -> DeclSyntax {
-            guard node.hasRedundantDiscardableLetViolation else {
-                return super.visit(node)
-            }
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        private var codeBlockScopes = Stack<CodeBlockKind>()
 
-            correctionPositions.append(node.positionAfterSkippingLeadingTrivia)
-            let newNode = node
-                .with(\.bindingSpecifier, .keyword(.let, presence: .missing))
-                .with(\.bindings, node.bindings.with(\.leadingTrivia, node.bindingSpecifier.leadingTrivia))
-            return super.visit(newNode)
+        override func visit(_ node: AccessorBlockSyntax) -> SyntaxVisitorContinueKind {
+            codeBlockScopes.push(node.isViewBody ? .view : .normal)
+            return .visitChildren
+        }
+
+        override func visitPost(_: AccessorBlockSyntax) {
+            codeBlockScopes.pop()
+        }
+
+        override func visit(_: CodeBlockSyntax) -> SyntaxVisitorContinueKind {
+            codeBlockScopes.push(.normal)
+            return .visitChildren
+        }
+
+        override func visitPost(_: CodeBlockSyntax) {
+            codeBlockScopes.pop()
+        }
+
+        override func visitPost(_ node: VariableDeclSyntax) {
+            if codeBlockScopes.peek() != .view || !configuration.ignoreSwiftUIViewBodies,
+               node.bindingSpecifier.tokenKind == .keyword(.let),
+               let binding = node.bindings.onlyElement,
+               binding.pattern.is(WildcardPatternSyntax.self),
+               binding.typeAnnotation == nil,
+               !node.modifiers.contains(where: { $0.name.text == "async" }) {
+                violations.append(
+                    ReasonedRuleViolation(
+                        position: node.bindingSpecifier.positionAfterSkippingLeadingTrivia,
+                        correction: .init(
+                            start: node.bindingSpecifier.positionAfterSkippingLeadingTrivia,
+                            end: binding.pattern.positionAfterSkippingLeadingTrivia,
+                            replacement: ""
+                        )
+                    )
+                )
+            }
         }
     }
 }
 
-private extension VariableDeclSyntax {
-    var hasRedundantDiscardableLetViolation: Bool {
-           bindingSpecifier.tokenKind == .keyword(.let)
-        && bindings.count == 1
-        && bindings.first!.pattern.is(WildcardPatternSyntax.self)
-        && bindings.first!.typeAnnotation == nil
-        && modifiers.contains(where: { $0.name.text == "async" }) != true
+private extension AccessorBlockSyntax {
+    var isViewBody: Bool {
+        if let binding = parent?.as(PatternBindingSyntax.self),
+           binding.pattern.as(IdentifierPatternSyntax.self)?.identifier.text == "body",
+           let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) {
+            return type.someOrAnySpecifier.text == "some"
+                && type.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
+                && binding.parent?.parent?.is(VariableDeclSyntax.self) == true
+        }
+        return false
     }
 }

--- a/Tests/IntegrationTests/default_rule_configurations.yml
+++ b/Tests/IntegrationTests/default_rule_configurations.yml
@@ -795,6 +795,7 @@ reduce_into:
     opt-in: true
 redundant_discardable_let:
   severity: warning
+  ignore_swiftui_view_bodies: false
   meta:
     opt-in: false
 redundant_nil_coalescing:

